<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.12.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>singler_classic_markers: Classic marker detection in SingleR</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">singler_classic_markers
   </div>
   <div id="projectbrief">Classic marker detection for SingleR</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.12.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="doc-content">
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Classic marker detection in SingleR </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="md__2github_2workspace_2README"></a></p>
<p><img src="https://github.com/SingleR-inc/singler_classic_markers/actions/workflows/run-tests.yaml/badge.svg" alt="Unit tests" style="pointer-events: none;" class="inline"/> <img src="https://github.com/SingleR-inc/singler_classic_markers/actions/workflows/doxygenate.yaml/badge.svg" alt="Documentation" style="pointer-events: none;" class="inline"/> <a href="https://codecov.io/gh/SingleR-inc/singler_classic_markers"><img src="https://codecov.io/gh/SingleR-inc/singler_classic_markers/branch/master/graph/badge.svg?token=q3AJjQtKtj" alt="codecov" style="pointer-events: none;" class="inline"/></a></p>
<h1>Overview</h1>
<p>This repository implements the classic marker detection algorithm from <a href="https://bioconductor.org/packages/SingleR"><b>SingleR</b></a>. For each pairwise comparison between labels, genes are ranked based on the difference in the medians of log-expression values. The top genes are then used as markers in the <b>SingleR</b> algorithm. We also provide methods for combining information across blocks in the presence of, e.g., batch effects.</p>
<h1>Quick start</h1>
<p><b><a class="el" href="namespacesingler__classic__markers.html" title="Implementation of the classic SingleR marker detection.">singler_classic_markers</a></b> is a header-only library, so it can be easily used by just <code>#include</code>ing the relevant source files.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &quot;<a class="code" href="singler__classic__markers_8hpp.html">singler_classic_markers/singler_classic_markers.hpp</a>&quot;</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// Prepare the reference matrix as a tatami::NumericMatrix.</span></div>
<div class="line">ref_mat;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Prepare a vector of labels, one per column of ref_mat.</span></div>
<div class="line">ref_labels;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Computing the top markers, which can be directly used in singlepp::train_single().</span></div>
<div class="line"><a class="code hl_struct" href="structsingler__classic__markers_1_1ChooseOptions.html">singler_classic_markers::ChooseOptions</a> opts;</div>
<div class="line"><span class="keyword">auto</span> markers = <a class="code hl_function" href="namespacesingler__classic__markers.html#a02dd6670d70dbc7a563785a95282d93b">singler_classic_markers::choose_index</a>(ref_mat, ref_labels.data(), opts); </div>
<div class="ttc" id="anamespacesingler__classic__markers_html_a02dd6670d70dbc7a563785a95282d93b"><div class="ttname"><a href="namespacesingler__classic__markers.html#a02dd6670d70dbc7a563785a95282d93b">singler_classic_markers::choose_index</a></div><div class="ttdeci">std::vector&lt; std::vector&lt; std::vector&lt; Index_ &gt; &gt; &gt; choose_index(const tatami::Matrix&lt; Value_, Index_ &gt; &amp;matrix, const Label_ *label, const ChooseOptions &amp;options)</div><div class="ttdef"><b>Definition</b> choose.hpp:153</div></div>
<div class="ttc" id="asingler__classic__markers_8hpp_html"><div class="ttname"><a href="singler__classic__markers_8hpp.html">singler_classic_markers.hpp</a></div><div class="ttdoc">Umbrella header for the singler_classic_markers library.</div></div>
<div class="ttc" id="astructsingler__classic__markers_1_1ChooseOptions_html"><div class="ttname"><a href="structsingler__classic__markers_1_1ChooseOptions.html">singler_classic_markers::ChooseOptions</a></div><div class="ttdoc">Options for choose().</div><div class="ttdef"><b>Definition</b> choose.hpp:26</div></div>
</div><!-- fragment --><p>See the <a href="https://singler-inc.github.io/singler_classic_markers">reference documentation</a> for more details.</p>
<h1>Alternatives</h1>
<p>As its name suggests, the classic method was part of the original implementation of the <b>SingleR</b> method. It is simple and works well for the built-in reference datasets, but is less appropriate for single-cell references where the median is often zero within each label. In such cases, we suggest using the <code>score_markers_best()</code> function from the <a href="https://github.com/libscran/scran_markers"><b>scran_markers</b></a> library, which ranks genes on effect sizes that are more robust to a large number of zeros.</p>
<p>For bulk RNA-seq references, we could also consider using tools like <a href="https://bioconductor.org/packages/edgeR"><b>edgeR</b></a> or <a href="https://bioconductor.org/packages/DESeq2"><b>DESeq2</b></a>. Briefly, we could perform differential expression comparisons between all labels (possibly using TREAT to enforce a minimum log-fold change) and then take the top genes as markers. This would likely yield better markers than <b>SingleR</b>'s classic approach, as <b>edgeR</b> and friends consider the variance within each label and will favor genes that have more consistent behavior.</p>
<h1>Building projects</h1>
<h2>CMake with <code>FetchContent</code></h2>
<p>If you're using CMake, you just need to add something like this to your <code>CMakeLists.txt</code>:</p>
<div class="fragment"><div class="line">include(FetchContent)</div>
<div class="line"> </div>
<div class="line">FetchContent_Declare(</div>
<div class="line">  singler_classic_markers</div>
<div class="line">  GIT_REPOSITORY https://github.com/SingleR-inc/singler_classic_markers</div>
<div class="line">  GIT_TAG master # replace with a pinned release</div>
<div class="line">)</div>
<div class="line"> </div>
<div class="line">FetchContent_MakeAvailable(singler_classic_markers)</div>
</div><!-- fragment --><p>Then you can link to <b><a class="el" href="namespacesingler__classic__markers.html" title="Implementation of the classic SingleR marker detection.">singler_classic_markers</a></b> to make the headers available during compilation:</p>
<div class="fragment"><div class="line"># For executables:</div>
<div class="line">target_link_libraries(myexe singler_classic_markers)</div>
<div class="line"> </div>
<div class="line"># For libaries</div>
<div class="line">target_link_libraries(mylib INTERFACE singler_classic_markers)</div>
</div><!-- fragment --><p>By default, this will use <code>FetchContent</code> to fetch all external dependencies. Applications are advised to pin the versions of all dependencies - see <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a> for suggested versions. If you want to install them manually, use <code>-DSINGLER_CLASSIC_MARKERS_FETCH_EXTERN=OFF</code>.</p>
<h2>CMake with <code>find_package()</code></h2>
<div class="fragment"><div class="line">find_package(singler_singler_classic_markers CONFIG REQUIRED)</div>
<div class="line">target_link_libraries(mylib INTERFACE singler::singler_classic_markers)</div>
</div><!-- fragment --><p>To install the library, use:</p>
<div class="fragment"><div class="line">mkdir build &amp;&amp; cd build</div>
<div class="line">cmake .. -DSINGLER_CLASSIC_MARKERS_TESTS=OFF</div>
<div class="line">cmake --build . --target install</div>
</div><!-- fragment --><p>Again, this will use <code>FetchContent</code> to retrieve dependencies, see comments above.</p>
<h2>Manual</h2>
<p>If you're not using CMake, the simple approach is to just copy the files in <code>include/</code> - either directly or with Git submodules - and include their path during compilation with, e.g., GCC's <code>-I</code>. This assumes that the external dependencies listed in <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a> are available during compilation.</p>
<h1>References</h1>
<p>Aran D et al. (2019). Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage. <em>Nat. Immunol.</em> 20, 163-172 </p>
</div></div><!-- PageDoc -->
<a href="doxygen_crawl.html"></a>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.12.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
